qpdf-convert-client.py

```diff
@@ -0,0 +1,257 @@
+                      size=untrusted_size, depth=depth)
+
+def recv_rgb_file(rgb_path, untrusted_size):
+    # XXX: For some reason, this leaves us missing alot of bytes
+    # rcvd_bytes = input().encode('utf-8', 'surrogateescape')
+    # rcvd_bytes = rcvd_bytes[:dimensions.size]
+
+    # XXX: Example of using PIL for performant PNG -> JPG. Maybe use this?
+    # png = Image.open(object.logo.path)
+    # png.load() # required for png.split()
+    # background = Image.new("RGB", png.size, (255, 255, 255))
+    # background.paste(png, mask=png.split()[3]) # 3 is the alpha channel
+    # background.save('foo.jpg', 'JPEG', quality=80)
+
+    with open(rgb_path, 'wb') as f:
+        # FIXME: Why doesn't this work in pure Python?
+        cmd = ['head', '-c', str(untrusted_size)]
+        subprocess.run(cmd, stdout=f, check=True)
+
+        if os.path.getsize(f.name) != untrusted_size:
+            os.remove(rgb_path)
+            die('Invalid number of bytes in RGB file... aborting!')
+
+def rgb_to_png(rgb_path, png_path, untrusted_dimensions, page):
+    cmd = ['convert', '-size',
+           f'{untrusted_dimensions.width}x{untrusted_dimensions.height}',
+           '-depth', str(untrusted_dimensions.depth), f'rgb:{rgb_path}',
+           f'png:{png_path}']
+
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        die(f'Page {page} conversion failed (RGB->PNG)... aborting!')
+    else:
+        os.remove(rgb_path)
+
+def convert_rgb_file(untrusted_dimensions, page):
+    with NamedTemporaryFile(prefix='qpdf-conversion-') as f:
+        rgb_path = f'{f.name}-{page}.rgb'
+        png_path = f'{f.name}-{page}.png'
+        recv_rgb_file(rgb_path, untrusted_dimensions.size)
+        rgb_to_png(rgb_path, png_path, untrusted_dimensions, page)
+
+    return png_path
+
+
+###############################
+#         PDF-related
+###############################
+
+def recv_page_count():
+    try:
+        untrusted_page_count = int(recvline_b().decode())
+        check_range(untrusted_page_count, MAX_PAGES)
+    except ValueError:
+        die("Invalid number of pages returned... aborting!")
+
+    return untrusted_page_count
+
+def send_pdf(untrusted_pdf_path):
+    info(f'Sending {untrusted_pdf_path} to a Disposable VM...')
+
+    # To process multiple files, we have to avoid closing STDIN since we can't
+    # reopen it afterwards without duplicating it to some new fd which doesn't
+    # seem ideal. Unfortunately, unless STDIN is being read from a terminal, I
+    # couldn't find a way to indicate to the server that we were done sending
+    # stuff.
+    #
+    # So, the current solution is to send file's size in advance so that the
+    # server can know when to stop reading from STDIN. The problem then becomes
+    # that the server may start its read after we send the PDF file.  Thus, we
+    # make the client sleep so that the server can start its read beforehand.
```

Comment:
> I think server reading just one line first should fix this issue.

This isn't an issue anymore in local repo, so we're all good there :+1:

> Anyway, this is protocol change, so it will break if only one side (client or server) is updated. Changes like this needs to take this into account. One solution is to change qrexec service name for the new protocol. And also handle old protocol under the old name (in this case it should be easy within the same code base).

Oh, good point.

I've actually wanted to have the service name be something like `Qubes.Sanitize` or `Qubes.DocSanitize` (and the program itself to `qvm-sanitize`). imo "convert-pdf" is a bit too specific and "convert" is just a bad word for describing this.

As for the old protocol, I really don't want to support it in the Python files themselves (if that's what you meant). I think a better option would be to keep the old shell scripts around so that `qvm-convert-pdf` will use the normal stuff while `qvm-sanitize` or something uses the new service and (of course, this is only until we can phase out the shell scripts and `qvm-convert-pdf`). This would only take minor changes to the Makefile and the other packaging files.

In any case, I think I should move this PR to another one and call it "Rewrite PDF Converter" or something. I feel bad for having all these changes in what's supposed to be a port lol.